common/matrix/csr_kernels.hpp.inc

```diff
@@ -494,95 +494,95 @@ __global__ __launch_bounds__(spmv_block_size) void abstract_classical_spmv(
 __global__ __launch_bounds__(default_block_size) void spgeam_nnz(
     const IndexType *a_row_ptrs, const IndexType *a_col_idxs,
     const IndexType *b_row_ptrs, const IndexType *b_col_idxs,
     IndexType num_rows, IndexType *nnz)
 {
-    constexpr auto sentinel = device_numeric_limits<IndexType>::max;
-    auto row = threadIdx.x + blockDim.x * static_cast<size_type>(blockIdx.x);
+    auto row =
+        (threadIdx.x + blockDim.x * size_type(blockIdx.x)) / subwarp_size;
+    auto subwarp =
+        group::tiled_partition<subwarp_size>(group::this_thread_block());
     if (row >= num_rows) {
         return;
     }
 
     auto a_begin = a_row_ptrs[row];
     auto b_begin = b_row_ptrs[row];
-    auto a_end = a_row_ptrs[row + 1];
-    auto b_end = b_row_ptrs[row + 1];
-    auto a_size = a_end - a_begin;
-    auto b_size = b_end - b_begin;
-    auto a_ptr = a_begin;
-    auto b_ptr = b_begin;
-    auto a_col = a_ptr < a_end ? a_col_idxs[a_ptr] : sentinel;
-    auto b_col = b_ptr < b_end ? b_col_idxs[b_ptr] : sentinel;
+    auto a_size = a_row_ptrs[row + 1] - a_begin;
+    auto b_size = b_row_ptrs[row + 1] - b_begin;
     IndexType count{};
-    IndexType i{};
-    while (i < a_size + b_size) {
-        auto advance_a = (a_col <= b_col);
-        auto advance_b = (b_col <= a_col);
-        i += advance_a + advance_b;
-        a_ptr += advance_a;
-        b_ptr += advance_b;
-        if (advance_a) {
-            a_col = a_ptr < a_end ? a_col_idxs[a_ptr] : sentinel;
-        }
-        if (advance_b) {
-            b_col = b_ptr < b_end ? b_col_idxs[b_ptr] : sentinel;
-        }
-        ++count;
+    group_merge<subwarp_size>(
+        a_col_idxs + a_begin, a_size, b_col_idxs + b_begin, b_size, subwarp,
+        [&](IndexType, IndexType a_col, IndexType, IndexType b_col, IndexType,
+            bool valid) {
+            count += popcnt(subwarp.ballot(a_col != b_col && valid));
+        });
+
+    if (subwarp.thread_rank() == 0) {
+        nnz[row] = count;
     }
-    nnz[row] = count;
 }
 
 
-template <typename IndexType, typename ValueType>
+template <int subwarp_size, typename IndexType, typename ValueType>
 __global__ __launch_bounds__(default_block_size) void spgeam(
     ValueType alpha, const IndexType *a_row_ptrs, const IndexType *a_col_idxs,
     const ValueType *a_vals, ValueType beta, const IndexType *b_row_ptrs,
     const IndexType *b_col_idxs, const ValueType *b_vals, IndexType num_rows,
     const IndexType *c_row_ptrs, IndexType *c_col_idxs, ValueType *c_vals)
 {
-    constexpr auto sentinel = device_numeric_limits<IndexType>::max;
-    auto row = threadIdx.x + blockDim.x * static_cast<size_type>(blockIdx.x);
+    auto row =
+        (threadIdx.x + blockDim.x * size_type(blockIdx.x)) / subwarp_size;
+    auto subwarp =
+        group::tiled_partition<subwarp_size>(group::this_thread_block());
     if (row >= num_rows) {
         return;
     }
 
+    auto lane = IndexType(subwarp.thread_rank());
```

Comment:
Why not simply:
```suggestion
    IndexType lane = subwarp.thread_rank();
```
instead of a functional cast? Also, I do consider functional- & C-casts dangerous if done with a template parameter (which is not guaranteed to be a simple type).

Reply:
Good point, that is probably easier. I am probably too heavily in the "[AAAARGH](https://www.youtube.com/watch?v=ZCGyvPDM0YY)" camp with my code. But I would seriously question every piece of code that used a non-primitive IndexType.

Reply:
Wait no, thread_rank(), as well as most other functions of cooperative groups, returns an unsigned value, so this would produce a warning

Reply:
I agree that IndexType should be a primitive type.
I guess there are edge cases where if you call a function with a wrong parameter type, it could happen that `IndexType` is suddenly a pointer or something completely different. I hope it would not compile that way, but I guess it could happen.